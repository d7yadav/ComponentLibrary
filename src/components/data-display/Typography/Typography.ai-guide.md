# Typography Component AI Guide

## ü§ñ AI Assistant Instructions

### Component Overview
The Typography component is a comprehensive text rendering system with 13 typography variants, advanced accessibility features, and enhanced styling capabilities. It serves as the foundation for all text display in the design system.

### Key Features for AI Implementation
- **13 Typography Variants**: h1-h6, body1-2, button, caption, overline, subtitle1-2
- **Enhanced Accessibility**: WCAG 2.1 AA compliance with ARIA support
- **Interactive Text Support**: Click handlers and button/link roles
- **Gradient Text**: Advanced gradient text rendering
- **Multi-line Truncation**: Configurable text truncation
- **Responsive Typography**: Automatic scaling across devices
- **Font Feature Settings**: Advanced typography features

## üéØ AI Usage Patterns

### 1. Basic Text Display
```typescript
// Simple text rendering
<Typography variant="body1">
  Standard body text content
</Typography>

// Heading with semantic HTML
<Typography variant="h2" component="h2">
  Section Heading
</Typography>
```

### 2. Enhanced Typography Features
```typescript
// Multi-line truncation
<Typography 
  variant="body1" 
  maxLines={3}
  title="Full text for accessibility"
>
  Long text content that will be truncated after 3 lines...
</Typography>

// Interactive text with click handling
<Typography 
  variant="button" 
  role="button"
  onClick={handleClick}
  sx={{ cursor: 'pointer' }}
>
  Clickable Text
</Typography>
```

### 3. Specialized Typography Components
```typescript
// Use pre-configured variants for common patterns
<HeadingTypography variant="h1">Main Title</HeadingTypography>
<BodyTypography>Standard paragraph text</BodyTypography>
<CaptionTypography>Secondary information</CaptionTypography>
<LabelTypography>Form label text</LabelTypography>
```

## üé® Styling Guidelines

### Color System Integration
```typescript
// Use theme colors for consistency
<Typography color="primary.main">Primary text</Typography>
<Typography color="text.secondary">Secondary text</Typography>
<Typography color="error.main">Error message</Typography>
```

### Responsive Typography
```typescript
// Enable responsive scaling (default: true)
<Typography 
  variant="h1" 
  responsive={true}
  sx={{
    fontSize: { xs: '2rem', md: '3rem', lg: '4rem' }
  }}
>
  Responsive Heading
</Typography>
```

### Gradient Text Effects
```typescript
// Advanced gradient text
<Typography 
  variant="h2"
  sx={{
    background: 'linear-gradient(45deg, #FF6B6B, #4ECDC4)',
    WebkitBackgroundClip: 'text',
    WebkitTextFillColor: 'transparent',
  }}
>
  Gradient Heading
</Typography>
```

## ‚ôø Accessibility Implementation

### ARIA Support
```typescript
// Proper heading hierarchy
<Typography 
  variant="h3" 
  component="h3"
  aria-level={3}
  id="section-heading"
>
  Section Title
</Typography>

// Descriptive content
<Typography 
  variant="body2"
  aria-describedby="help-text"
  aria-label="Important notice"
>
  Content with accessibility enhancements
</Typography>
```

### Screen Reader Optimization
```typescript
// Content that may be truncated
<Typography 
  maxLines={2}
  title="Full content for screen readers"
  aria-label="Summary text, full content available on interaction"
>
  Truncated content...
</Typography>
```

## üß© Common AI Patterns

### 1. Content Cards
```typescript
<Card>
  <CardContent>
    <HeadingTypography variant="h5" gutterBottom>
      Card Title
    </HeadingTypography>
    <BodyTypography color="text.secondary">
      Card description content
    </BodyTypography>
    <CaptionTypography sx={{ mt: 2 }}>
      Last updated: {date}
    </CaptionTypography>
  </CardContent>
</Card>
```

### 2. Form Labels and Help Text
```typescript
<Box>
  <LabelTypography htmlFor="email-input">
    Email Address *
  </LabelTypography>
  <TextField id="email-input" />
  <CaptionTypography color="text.secondary">
    We'll never share your email address
  </CaptionTypography>
</Box>
```

### 3. Status Messages
```typescript
// Error state
<Typography 
  variant="body2" 
  color="error.main"
  role="alert"
  aria-live="polite"
>
  Validation error message
</Typography>

// Success state
<Typography 
  variant="body2" 
  color="success.main"
  role="status"
  aria-live="polite"
>
  Operation completed successfully
</Typography>
```

## üé≠ Advanced Use Cases

### Custom Typography Variants
```typescript
// Create custom variants using sx prop
<Typography
  variant="body1"
  sx={{
    fontFamily: 'monospace',
    backgroundColor: 'grey.100',
    padding: 1,
    borderRadius: 1,
    fontSize: '0.875rem',
  }}
>
  Code snippet text
</Typography>
```

### Dynamic Content Handling
```typescript
// Handle dynamic content with proper accessibility
<Typography
  variant={isImportant ? 'h6' : 'body2'}
  color={isError ? 'error.main' : 'text.primary'}
  aria-live={isError ? 'assertive' : 'polite'}
  role={isError ? 'alert' : undefined}
>
  {dynamicMessage}
</Typography>
```

### Performance Optimization
```typescript
// Use specialized components for better performance
const MemoizedHeading = memo(() => (
  <HeadingTypography variant="h2">
    {expensiveComputedTitle}
  </HeadingTypography>
));
```

## üö® Common Pitfalls to Avoid

### ‚ùå Incorrect Usage
```typescript
// Don't use div when semantic HTML is needed
<Typography variant="h1" component="div">Heading</Typography>

// Don't ignore accessibility for interactive text
<Typography onClick={handleClick}>Clickable</Typography>

// Don't use hardcoded colors
<Typography sx={{ color: '#ff0000' }}>Red text</Typography>
```

### ‚úÖ Correct Usage
```typescript
// Use semantic HTML elements
<Typography variant="h1" component="h1">Heading</Typography>

// Proper interactive text with accessibility
<Typography 
  role="button" 
  onClick={handleClick}
  onKeyDown={handleKeyDown}
  tabIndex={0}
  sx={{ cursor: 'pointer' }}
>
  Clickable
</Typography>

// Use theme colors
<Typography color="error.main">Error text</Typography>
```

## üìä Performance Considerations

1. **Component Selection**: Use specialized components (HeadingTypography, BodyTypography) when appropriate
2. **Memoization**: The component is already memoized, but consider memoizing parent components with expensive computations
3. **Font Features**: Advanced font features are enabled by default for better readability
4. **Responsive Scaling**: Uses CSS clamp() for optimal performance across devices

## üîç Testing Recommendations

```typescript
// Test accessibility features
expect(screen.getByRole('heading', { level: 2 })).toBeInTheDocument();
expect(screen.getByLabelText('Important notice')).toBeInTheDocument();

// Test interactive functionality
fireEvent.click(screen.getByRole('button'));
expect(mockClickHandler).toHaveBeenCalled();

// Test truncation
expect(container.querySelector('[title]')).toHaveAttribute('title', 'Full text content');
```

## üéØ AI Decision Tree

**When to use Typography variants:**
- `h1-h6`: Semantic headings with proper hierarchy
- `body1-body2`: Standard paragraph content (body1 is larger)
- `button`: Interactive text elements
- `caption`: Secondary information, metadata
- `overline`: Category labels, small headers
- `subtitle1-subtitle2`: Section subheadings

**When to use specialized components:**
- `HeadingTypography`: Quick heading with sensible defaults
- `BodyTypography`: Paragraph text with paragraph spacing
- `CaptionTypography`: Small secondary text
- `LabelTypography`: Form labels and UI labels

This component provides comprehensive typography solutions while maintaining accessibility and performance standards.