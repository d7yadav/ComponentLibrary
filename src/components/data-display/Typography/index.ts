export { 
  Typography,
  HeadingTypography,
  BodyTypography,
  CaptionTypography,
  LabelTypography
} from './Typography';

export type { 
  TypographyProps,
  TypographyVariant,
  TypographyColor,
  TypographyAlign,
  TypographyTransform,
  TypographyWeight,
  TypographyDisplay,
  TypographyComponent,
  TypographyStyleProps,
  ResponsiveTypographyConfig
} from './Typography.types';

export {
  TYPOGRAPHY_VARIANTS,
  TYPOGRAPHY_COLORS,
  TYPOGRAPHY_ALIGNMENTS,
  TYPOGRAPHY_TRANSFORMS,
  TYPOGRAPHY_WEIGHTS,
  TYPOGRAPHY_DISPLAYS,
  TYPOGRAPHY_COMPONENTS,
  TYPOGRAPHY_VARIANT_MAPPINGS,
  DEFAULT_PROPS as TYPOGRAPHY_DEFAULT_PROPS,
  ACCESSIBILITY_CONSTANTS as TYPOGRAPHY_ACCESSIBILITY_CONSTANTS,
  GRADIENT_PRESETS as TYPOGRAPHY_GRADIENT_PRESETS,
  FONT_FEATURE_SETTINGS
} from './Typography.constants';