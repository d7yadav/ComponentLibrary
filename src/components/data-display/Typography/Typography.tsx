import React, { forwardRef, memo } from 'react';
import { TypographyProps } from './Typography.types';
import { 
  StyledTypography, 
  InteractiveTypography, 
  GradientTypography, 
  AccessibleTypography 
} from './Typography.styles';
import { 
  TYPOGRAPHY_VARIANTS,
  TYPOGRAPHY_COLORS,
  TYPOGRAPHY_ALIGNMENTS,
  TYPOGRAPHY_TRANSFORMS,
  TYPOGRAPHY_WEIGHTS,
  TYPOGRAPHY_DISPLAYS,
  TYPOGRAPHY_VARIANT_MAPPINGS,
  DEFAULT_PROPS,
  ACCESSIBILITY_CONSTANTS,
  FONT_FEATURE_SETTINGS
} from './Typography.constants';

/**
 * Enhanced Typography component with comprehensive variant support and accessibility features
 * 
 * Features:
 * - 13 typography variants (h1-h6, body1-2, button, caption, overline, subtitle1-2)
 * - Multiple color variants with theme integration
 * - Text alignment and transformation options
 * - Responsive typography scaling
 * - Multi-line truncation support
 * - Interactive states for clickable text
 * - Gradient text support
 * - Enhanced accessibility compliance (WCAG 2.1 AA)
 * - Font feature settings for advanced typography
 * - Theme integration with CSS variables
 */
const TypographyComponent = forwardRef<HTMLElement, TypographyProps>(({
  variant = DEFAULT_PROPS.variant,
  color = DEFAULT_PROPS.color,
  align = DEFAULT_PROPS.align,
  textTransform = DEFAULT_PROPS.textTransform,
  fontWeight = DEFAULT_PROPS.fontWeight,
  display = DEFAULT_PROPS.display,
  component,
  noWrap = DEFAULT_PROPS.noWrap,
  paragraph = DEFAULT_PROPS.paragraph,
  gutterBottom = DEFAULT_PROPS.gutterBottom,
  maxLines,
  responsive = DEFAULT_PROPS.responsive,
  fontSize,
  lineHeight,
  letterSpacing,
  children,
  className,
  sx,
  'aria-label': ariaLabel,
  'aria-describedby': ariaDescribedBy,
  'aria-level': ariaLevel,
  role,
  id,
  title,
  ...other
}, ref) => {
  // Determine the HTML element to render
  const elementComponent = component || TYPOGRAPHY_VARIANT_MAPPINGS[variant] || 'p';
  
  // Enhanced accessibility props
  const getAccessibilityProps = () => {
    const isHeading = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'].includes(variant);
    const props: Record<string, any> = {};
    
    // ARIA label
    if (ariaLabel) {
      props['aria-label'] = ariaLabel;
    }
    
    // ARIA described by
    if (ariaDescribedBy) {
      props['aria-describedby'] = ariaDescribedBy;
    }
    
    // ARIA level for headings
    if (isHeading && ariaLevel) {
      props['aria-level'] = ariaLevel;
    }
    
    // Role
    if (role) {
      props.role = role;
    }
    
    // Enhanced screen reader support
    if (maxLines && maxLines > 1) {
      props['aria-label'] = props['aria-label'] || `Text content, may be truncated to ${maxLines} lines`;
    }
    
    return props;
  };
  
  // Style props for the styled component
  const styleProps = {
    customVariant: variant,
    customColor: color,
    customAlign: align,
    textTransform,
    fontWeight,
    display,
    noWrap,
    paragraph,
    gutterBottom,
    maxLines,
    responsive,
    fontSize,
    lineHeight,
    letterSpacing,
  };
  
  // Enhanced sx prop with font features and accessibility
  const enhancedSx = {
    // Advanced font features for better typography
    fontFeatureSettings: FONT_FEATURE_SETTINGS.readability,
    
    // Ensure proper contrast in high contrast mode
    '@media (prefers-contrast: high)': {
      fontWeight: fontWeight === 'light' ? 'regular' : fontWeight,
    },
    
    // Respect user's motion preferences
    '@media (prefers-reduced-motion: reduce)': {
      transition: 'none',
    },
    
    // Custom sx overrides
    ...sx,
  };
  
  // Base component props
  const baseProps = {
    ref,
    component: elementComponent,
    variant: variant as any, // MUI typing compatibility
    className,
    sx: enhancedSx,
    id,
    title,
    ...getAccessibilityProps(),
    ...styleProps,
    ...other,
  };
  
  // Determine which styled component to use based on enhanced features
  const shouldUseGradient = sx?.background && (sx.background.includes('gradient') || sx.background.includes('linear-gradient'));
  const shouldUseInteractive = Boolean(other.onClick || role === 'button' || role === 'link');
  const shouldUseAccessible = color.includes('text.') || variant.startsWith('h') || maxLines;
  
  // Render appropriate styled component variant
  if (shouldUseGradient) {
    return (
      <GradientTypography data-testid="typography"
        {...baseProps}
        gradient={sx?.background as string}
      >
        {children}
      </GradientTypography>
    );
  }
  
  if (shouldUseInteractive) {
    return (
      <InteractiveTypography
        {...baseProps}
        interactive={true}
      >
        {children}
      </InteractiveTypography>
    );
  }
  
  if (shouldUseAccessible) {
    return (
      <AccessibleTypography {...baseProps}>
        {children}
      </AccessibleTypography>
    );
  }
  
  // Default styled typography
  return (
    <StyledTypography {...baseProps}>
      {children}
    </StyledTypography>
  );
});

TypographyComponent.displayName = 'Typography';

// Export memoized component for performance optimization
export const Typography = memo(TypographyComponent);

// Export additional utility components for specific use cases
export const HeadingTypography = memo(forwardRef<HTMLElement, TypographyProps>((props, ref) => (
  <Typography
    {...props}
    ref={ref}
    variant={props.variant || 'h2'}
    component={props.component || 'h2'}
    gutterBottom={props.gutterBottom ?? true}
    fontWeight={props.fontWeight || 'medium'}
  />
)));

HeadingTypography.displayName = 'HeadingTypography';

export const BodyTypography = memo(forwardRef<HTMLElement, TypographyProps>((props, ref) => (
  <Typography
    {...props}
    ref={ref}
    variant={props.variant || 'body1'}
    component={props.component || 'p'}
    paragraph={props.paragraph ?? true}
  />
)));

BodyTypography.displayName = 'BodyTypography';

export const CaptionTypography = memo(forwardRef<HTMLElement, TypographyProps>((props, ref) => (
  <Typography
    {...props}
    ref={ref}
    variant="caption"
    color={props.color || 'text.secondary'}
    component={props.component || 'span'}
  />
)));

CaptionTypography.displayName = 'CaptionTypography';

export const LabelTypography = memo(forwardRef<HTMLElement, TypographyProps>((props, ref) => (
  <Typography
    {...props}
    ref={ref}
    variant="button"
    component={props.component || 'label'}
    fontWeight={props.fontWeight || 'medium'}
  />
)));

LabelTypography.displayName = 'LabelTypography';