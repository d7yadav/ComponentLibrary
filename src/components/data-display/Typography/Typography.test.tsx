/**
 * @fileoverview Typography Component Tests
 * @author AI Generated Tests - 2025-07-13
 */

import { ThemeProvider , createTheme } from '@mui/material/styles';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe, toHaveNoViolations } from 'jest-axe';

import { Typography } from './Typography';
import type { TypographyProps } from './Typography.types';

// Add jest-axe matcher
expect.extend(toHaveNoViolations);

// Test theme
const testTheme = createTheme();

// Test wrapper component
const TestWrapper = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={testTheme}>{children}</ThemeProvider>
);

// Helper function to render component with theme
const renderWithTheme = (ui: React.ReactElement, options = {}) => {
  return render(ui, { wrapper: TestWrapper, ...options });
};

describe('Typography Component', () => {
  // Basic rendering tests
  describe('Rendering', () => {
    it('renders without crashing', () => {
      renderWithTheme(<Typography>Test content</Typography>);
      expect(screen.getByText('Test content')).toBeInTheDocument();
    });

    it('renders children correctly', () => {
      const testContent = 'Test Typography Content';
      renderWithTheme(<Typography>{testContent}</Typography>);
      expect(screen.getByText(testContent)).toBeInTheDocument();
    });

    it('renders with custom test id', () => {
      const testId = 'custom-typography';
      renderWithTheme(
        <Typography data-testid={testId}>Test</Typography>
      );
      expect(screen.getByTestId(testId)).toBeInTheDocument();
    });

    it('applies custom className', () => {
      const customClass = 'custom-class';
      renderWithTheme(
        <Typography className={customClass} data-testid="typography">Test</Typography>
      );
      expect(screen.getByTestId('typography')).toHaveClass(customClass);
    });
  });

  // Props testing
  describe('Props', () => {
    describe('variant prop', () => {
      const variants: TypographyProps['variant'][] = [
        'h1', 'h2', 'h3', 'h4', 'h5', 'h6',
        'body1', 'body2', 'button', 'caption', 'overline',
        'subtitle1', 'subtitle2'
      ];

      variants.forEach((variant) => {
        it(`renders ${variant} variant correctly`, () => {
          renderWithTheme(
            <Typography variant={variant} data-testid="variant-test">
              {variant} variant
            </Typography>
          );
          const element = screen.getByTestId('variant-test');
          expect(element).toBeInTheDocument();
        });
      });
    });

    describe('color prop', () => {
      const colors: TypographyProps['color'][] = [
        'primary', 'secondary', 'tertiary', 'quaternary',
        'success', 'warning', 'error', 'info',
        'text.primary', 'text.secondary', 'text.disabled'
      ];

      colors.forEach((color) => {
        it(`renders ${color} color correctly`, () => {
          renderWithTheme(
            <Typography color={color} data-testid="color-test">
              {color} color
            </Typography>
          );
          const element = screen.getByTestId('color-test');
          expect(element).toBeInTheDocument();
        });
      });
    });

    describe('align prop', () => {
      const alignments: TypographyProps['align'][] = [
        'left', 'center', 'right', 'justify'
      ];

      alignments.forEach((align) => {
        it(`renders ${align} alignment correctly`, () => {
          renderWithTheme(
            <Typography align={align} data-testid="align-test">
              {align} alignment
            </Typography>
          );
          const element = screen.getByTestId('align-test');
          expect(element).toBeInTheDocument();
        });
      });
    });

    it('handles noWrap correctly', () => {
      renderWithTheme(
        <Typography noWrap data-testid="nowrap-test">
          This is a long text that should not wrap and should be truncated with ellipsis
        </Typography>
      );
      const element = screen.getByTestId('nowrap-test');
      expect(element).toBeInTheDocument();
    });

    it('handles paragraph correctly', () => {
      renderWithTheme(
        <Typography paragraph data-testid="paragraph-test">
          Paragraph text
        </Typography>
      );
      const element = screen.getByTestId('paragraph-test');
      expect(element).toBeInTheDocument();
    });

    it('handles gutterBottom correctly', () => {
      renderWithTheme(
        <Typography gutterBottom data-testid="gutter-test">
          Text with gutter bottom
        </Typography>
      );
      const element = screen.getByTestId('gutter-test');
      expect(element).toBeInTheDocument();
    });

    it('handles maxLines correctly', () => {
      renderWithTheme(
        <Typography maxLines={2} data-testid="maxlines-test">
          This is a very long text that should be truncated after two lines when maxLines prop is set to 2
        </Typography>
      );
      const element = screen.getByTestId('maxlines-test');
      expect(element).toBeInTheDocument();
    });

    it('handles interactive correctly', () => {
      const handleClick = vi.fn();
      renderWithTheme(
        <Typography interactive onClick={handleClick} data-testid="interactive-test">
          Interactive text
        </Typography>
      );
      const element = screen.getByTestId('interactive-test');
      fireEvent.click(element);
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('handles gradient correctly', () => {
      renderWithTheme(
        <Typography gradient="primary" data-testid="gradient-test">
          Gradient text
        </Typography>
      );
      const element = screen.getByTestId('gradient-test');
      expect(element).toBeInTheDocument();
    });
  });

  // Component prop testing
  describe('Component Prop', () => {
    it('renders as different HTML elements', () => {
      const { rerender } = renderWithTheme(
        <Typography component="h1" data-testid="component-test">
          Header 1
        </Typography>
      );
      
      let element = screen.getByTestId('component-test');
      expect(element.tagName).toBe('H1');

      rerender(
        <TestWrapper>
          <Typography component="span" data-testid="component-test">
            Span element
          </Typography>
        </TestWrapper>
      );
      
      element = screen.getByTestId('component-test');
      expect(element.tagName).toBe('SPAN');
    });

    it('renders as custom component', () => {
      const CustomComponent = ({ children, ...props }: any) => (
        <div {...props} data-custom="true">{children}</div>
      );

      renderWithTheme(
        <Typography component={CustomComponent} data-testid="custom-component">
          Custom component
        </Typography>
      );
      
      const element = screen.getByTestId('custom-component');
      expect(element).toHaveAttribute('data-custom', 'true');
    });
  });

  // Interaction testing
  describe('Interactions', () => {
    it('handles click events when interactive', async () => {
      const user = userEvent.setup();
      const handleClick = vi.fn();
      
      renderWithTheme(
        <Typography interactive onClick={handleClick} data-testid="clickable">
          Clickable text
        </Typography>
      );
      
      const element = screen.getByTestId('clickable');
      await user.click(element);
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('handles keyboard navigation when interactive', async () => {
      const user = userEvent.setup();
      const handleClick = vi.fn();
      
      renderWithTheme(
        <Typography interactive onClick={handleClick} tabIndex={0} data-testid="keyboard-nav">
          Keyboard navigable text
        </Typography>
      );
      
      const element = screen.getByTestId('keyboard-nav');
      
      // Test that the element exists and is interactive
      expect(element).toBeInTheDocument();
      expect(element).toHaveAttribute('tabIndex', '0');
      
      // Test keyboard interaction
      await user.click(element);
      await user.keyboard('{Enter}');
      expect(handleClick).toHaveBeenCalled();
    });

    it('handles hover states when interactive', () => {
      renderWithTheme(
        <Typography interactive data-testid="hoverable">
          Hoverable text
        </Typography>
      );
      
      const element = screen.getByTestId('hoverable');
      fireEvent.mouseEnter(element);
      expect(element).toBeInTheDocument();
      
      fireEvent.mouseLeave(element);
      expect(element).toBeInTheDocument();
    });
  });

  // Accessibility testing
  describe('Accessibility', () => {
    it('meets WCAG accessibility guidelines', async () => {
      const { container } = renderWithTheme(
        <Typography>Accessible Typography</Typography>
      );
      
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('supports screen readers with proper semantics', () => {
      renderWithTheme(
        <Typography variant="h1" data-testid="heading">
          Main Heading
        </Typography>
      );
      
      const element = screen.getByRole('heading', { level: 1 });
      expect(element).toBeInTheDocument();
      expect(element).toHaveTextContent('Main Heading');
    });

    it('provides proper ARIA attributes when interactive', () => {
      renderWithTheme(
        <Typography 
          interactive
          aria-label="Interactive text"
          aria-describedby="description"
          data-testid="aria-text"
        >
          ARIA Typography
        </Typography>
      );
      
      const element = screen.getByTestId('aria-text');
      expect(element).toHaveAttribute('aria-label', 'Interactive text');
      expect(element).toHaveAttribute('aria-describedby', 'description');
    });

    it('supports keyboard navigation', () => {
      renderWithTheme(
        <Typography interactive tabIndex={0} data-testid="keyboard-typography">
          Keyboard accessible text
        </Typography>
      );
      
      const element = screen.getByTestId('keyboard-typography');
      element.focus();
      expect(element).toHaveFocus();
    });
  });

  // Theme integration testing
  describe('Theme Integration', () => {
    it('respects theme customizations', () => {
      const customTheme = createTheme({
        typography: {
          h1: {
            fontSize: '3rem',
          },
        },
      });

      render(
        <ThemeProvider theme={customTheme}>
          <Typography variant="h1" data-testid="themed-typography">
            Themed Typography
          </Typography>
        </ThemeProvider>
      );

      expect(screen.getByTestId('themed-typography')).toBeInTheDocument();
    });

    it('supports dark mode', () => {
      const darkTheme = createTheme({
        palette: {
          mode: 'dark',
        },
      });

      render(
        <ThemeProvider theme={darkTheme}>
          <Typography data-testid="dark-typography">
            Dark Mode Typography
          </Typography>
        </ThemeProvider>
      );

      expect(screen.getByTestId('dark-typography')).toBeInTheDocument();
    });
  });

  // Performance testing
  describe('Performance', () => {
    it('renders efficiently with many instances', () => {
      const manyTypographies = Array.from({ length: 100 }, (_, i) => (
        <Typography key={i} variant="body1">
          Typography {i}
        </Typography>
      ));

      const { container } = renderWithTheme(<div>{manyTypographies}</div>);
      expect(container.querySelectorAll('[data-testid]')).toHaveLength(0);
      // The Typography components should render without issues
      expect(container).toBeInTheDocument();
    });

    it('memoizes properly to prevent unnecessary re-renders', () => {
      const { rerender } = renderWithTheme(
        <Typography variant="h1" data-testid="memo-test">
          Memoized Typography
        </Typography>
      );

      const element = screen.getByTestId('memo-test');
      expect(element).toBeInTheDocument();

      // Re-render with same props should not cause issues
      rerender(
        <TestWrapper>
          <Typography variant="h1" data-testid="memo-test">
            Memoized Typography
          </Typography>
        </TestWrapper>
      );

      expect(screen.getByTestId('memo-test')).toBeInTheDocument();
    });
  });

  // Edge cases and error handling
  describe('Edge Cases', () => {
    it('handles empty children gracefully', () => {
      renderWithTheme(<Typography data-testid="empty-typography" />);
      expect(screen.getByTestId('empty-typography')).toBeInTheDocument();
    });

    it('handles null children gracefully', () => {
      renderWithTheme(
        <Typography data-testid="null-typography">{null}</Typography>
      );
      expect(screen.getByTestId('null-typography')).toBeInTheDocument();
    });

    it('handles undefined props gracefully', () => {
      renderWithTheme(
        <Typography 
          variant={undefined as any}
          color={undefined as any}
          data-testid="undefined-props"
        >
          Undefined Props
        </Typography>
      );
      expect(screen.getByTestId('undefined-props')).toBeInTheDocument();
    });

    it('handles very long text content', () => {
      const longText = 'A'.repeat(10000);
      renderWithTheme(
        <Typography data-testid="long-text">{longText}</Typography>
      );
      expect(screen.getByTestId('long-text')).toBeInTheDocument();
    });

    it('handles special characters and HTML entities', () => {
      const specialText = '&lt;script&gt;alert("xss")&lt;/script&gt; & special chars: àáâãäå';
      renderWithTheme(
        <Typography data-testid="special-chars">{specialText}</Typography>
      );
      expect(screen.getByTestId('special-chars')).toHaveTextContent(specialText);
    });
  });

  // Snapshots
  describe('Snapshots', () => {
    it('matches snapshot for default props', () => {
      const { container } = renderWithTheme(
        <Typography>Default Typography</Typography>
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for all variants', () => {
      const variants: TypographyProps['variant'][] = [
        'h1', 'h2', 'h3', 'h4', 'h5', 'h6',
        'body1', 'body2', 'button', 'caption', 'overline',
        'subtitle1', 'subtitle2'
      ];

      variants.forEach((variant) => {
        const { container } = renderWithTheme(
          <Typography variant={variant}>{variant} Typography</Typography>
        );
        expect(container.firstChild).toMatchSnapshot(`variant-${variant}`);
      });
    });
  });
});