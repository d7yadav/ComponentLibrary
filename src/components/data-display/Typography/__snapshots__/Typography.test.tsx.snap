// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Typography Component > Snapshots > matches snapshot for all variants > variant-body1 1`] = `
<p
  class="MuiTypography-root MuiTypography-body1 css-14n44wn-MuiTypography-root"
  customalign="inherit"
  customcolor="text.primary"
  customvariant="body1"
>
  body1
   Typography
</p>
`;

exports[`Typography Component > Snapshots > matches snapshot for all variants > variant-body2 1`] = `
<p
  class="MuiTypography-root MuiTypography-body2 css-1gwjj97-MuiTypography-root"
  customalign="inherit"
  customcolor="text.primary"
  customvariant="body2"
>
  body2
   Typography
</p>
`;

exports[`Typography Component > Snapshots > matches snapshot for all variants > variant-button 1`] = `
<span
  class="MuiTypography-root MuiTypography-button css-whzwq7-MuiTypography-root"
  customalign="inherit"
  customcolor="text.primary"
  customvariant="button"
>
  button
   Typography
</span>
`;

exports[`Typography Component > Snapshots > matches snapshot for all variants > variant-caption 1`] = `
<span
  class="MuiTypography-root MuiTypography-caption css-nswiet-MuiTypography-root"
  customalign="inherit"
  customcolor="text.primary"
  customvariant="caption"
>
  caption
   Typography
</span>
`;

exports[`Typography Component > Snapshots > matches snapshot for all variants > variant-h1 1`] = `
<h1
  class="MuiTypography-root MuiTypography-h1 css-13bxxjs-MuiTypography-root"
  customalign="inherit"
  customcolor="text.primary"
  customvariant="h1"
>
  h1
   Typography
</h1>
`;

exports[`Typography Component > Snapshots > matches snapshot for all variants > variant-h2 1`] = `
<h2
  class="MuiTypography-root MuiTypography-h2 css-1my1ga1-MuiTypography-root"
  customalign="inherit"
  customcolor="text.primary"
  customvariant="h2"
>
  h2
   Typography
</h2>
`;

exports[`Typography Component > Snapshots > matches snapshot for all variants > variant-h3 1`] = `
<h3
  class="MuiTypography-root MuiTypography-h3 css-1wsxnxh-MuiTypography-root"
  customalign="inherit"
  customcolor="text.primary"
  customvariant="h3"
>
  h3
   Typography
</h3>
`;

exports[`Typography Component > Snapshots > matches snapshot for all variants > variant-h4 1`] = `
<h4
  class="MuiTypography-root MuiTypography-h4 css-gwdu49-MuiTypography-root"
  customalign="inherit"
  customcolor="text.primary"
  customvariant="h4"
>
  h4
   Typography
</h4>
`;

exports[`Typography Component > Snapshots > matches snapshot for all variants > variant-h5 1`] = `
<h5
  class="MuiTypography-root MuiTypography-h5 css-5akx2e-MuiTypography-root"
  customalign="inherit"
  customcolor="text.primary"
  customvariant="h5"
>
  h5
   Typography
</h5>
`;

exports[`Typography Component > Snapshots > matches snapshot for all variants > variant-h6 1`] = `
<h6
  class="MuiTypography-root MuiTypography-h6 css-ru79qb-MuiTypography-root"
  customalign="inherit"
  customcolor="text.primary"
  customvariant="h6"
>
  h6
   Typography
</h6>
`;

exports[`Typography Component > Snapshots > matches snapshot for all variants > variant-overline 1`] = `
<span
  class="MuiTypography-root MuiTypography-overline css-1s2pfyc-MuiTypography-root"
  customalign="inherit"
  customcolor="text.primary"
  customvariant="overline"
>
  overline
   Typography
</span>
`;

exports[`Typography Component > Snapshots > matches snapshot for all variants > variant-subtitle1 1`] = `
<h6
  class="MuiTypography-root MuiTypography-subtitle1 css-coy9ar-MuiTypography-root"
  customalign="inherit"
  customcolor="text.primary"
  customvariant="subtitle1"
>
  subtitle1
   Typography
</h6>
`;

exports[`Typography Component > Snapshots > matches snapshot for all variants > variant-subtitle2 1`] = `
<h6
  class="MuiTypography-root MuiTypography-subtitle2 css-1wc4vlz-MuiTypography-root"
  customalign="inherit"
  customcolor="text.primary"
  customvariant="subtitle2"
>
  subtitle2
   Typography
</h6>
`;

exports[`Typography Component > Snapshots > matches snapshot for default props 1`] = `
<p
  class="MuiTypography-root MuiTypography-body1 css-14n44wn-MuiTypography-root"
  customalign="inherit"
  customcolor="text.primary"
  customvariant="body1"
>
  Default Typography
</p>
`;
